### ISSUE TITLE
How to separate "methods" and "plugins"

### ISSUE THREAD

**Issue by jacomyal** [medialab/graphlib#3]
The documentation exposes `Graph#connectedComponents()` as a method, but `louvain` as a dependant and separated library.

How to decide wether some feature should be implemented as a method or a plugin?

**Comment by Yomguithereal** [comment]
This is clearly one of the hardest point to define. Where do we draw the line? I would say in this very case that `connectedComponents` is solved in 10 lines whereas `louvain` is not :D

**Comment by DavidBruant** [comment]
> How to decide wether some feature should be implemented as a method or a plugin?

My point of view on the matter would be to have a very minimal API just to create (add/removeNode/Edge) and inspect the graph (getNodes/Edges) and move everything else to different modules. ngraph sets an interesting precedent in that direction.
In that case, both `louvain` and `connectedComponents` would be both separate modules (even if the latter is only 10 lines).

**Comment by Yomguithereal** [comment]
Properties' indexes (#4) is a good example of thing we should decide in or out.

**Comment by jacomyal** [comment]
> My point of view on the matter would be to have a very minimal API

Agreed. In most cases, when playing with sigma, I only needed the data manipulation / accessors of the graph model, and no actual graph algorithm. So, I'd prefer this lib to be as light as possible, and maybe some (community curated?) [marketplace](https://marketplace.gephi.org/plugins/) to share all types of graph algorithms as plugins.

Though, my initial question was also about the development pattern: When do we want to extend the graph prototype to add features rather than exposing functions that takes the graph as argument?

```javascript
// Basically, do you prefer this:
const components = graph.connectedComponents();
const communities = graph.louvain();

// Or this:
const components = connectedComponents(graph);
const communities = louvain(graph);
```

**Comment by Yomguithereal** [comment]
> So, I'd prefer this lib to be as light as possible, and maybe some (community curated?) marketplace to share all types of graph algorithms as plugins.

Yes, we should go towards this. Simply with prefixed packages like `{name}-louvain`.

I guess the second pattern is better however since:

1. One would not extend the prototype of a native object.
2. It keeps the possibility to extend the prototype very easily and thus keeps both methods possible:

```js
Graph.prototype.louvain = function() {
  return louvain(this);
});
```

**Comment by DavidBruant** [comment]
Another option that hasn't been mentioned in this thread is ES6 `class` + `extends`

````js
import Graph from 'graph';
import connectedComponents from 'graph-connected-components';
import louvain from 'graph-louvain';

class LouvainConnComDegreeGraph extends Graph(){
    louvain(){
        return louvain(this)
    },
    connectedComponents(){
        return connectedComponents(this)
    },
    degree(node){
        // compute degree for node
    }
}
````

This way, no one touches the initial graph, they interact only through the API, but they can still easily create a specialized graph that allows them to do `myGraph.louvain()`. Best of all worlds?

**Comment by Yomguithereal** [comment]
Not bad indeed.
